crates/red_knot_python_semantic/src/symbol.rs

```diff
@@ -667,9 +677,33 @@ fn symbol_from_bindings_impl<'db>(
         |BindingWithConstraints {
              binding,
              narrowing_constraint,
-             visibility_constraint,
+             mut visibility_constraint,
          }| {
-            let binding = binding?;
+            let mut binding = binding?;
+
+            if binding.kind(db).is_star_import() {
+                let use_def_map = use_def_map(db, scope);
+
+                while let Some(star_import) = binding.kind(db).as_star_import() {
+                    if resolve_star_import_definition(
+                        db,
+                        scope.file(db),
+                        star_import,
+                        &symbol_table(db, scope),
+                    )
+                    .ok()
+                    .is_none_or(|symbol| symbol.symbol.is_unbound())
+                    {
+                        let prior_binding = use_def_map
+                            .public_bindings_prior_to_star_import(binding)
+                            .next()?;
```

Comment:
calling `.next()` here and only considering the _first_ definition of the definitions-before-the-`*`-import is just... obviously incorrect :-)

but I'm not sure what the correct solution is here

Reply:
I think the simplest approach here might be to make a recursive `symbol_from_bindings_impl` call here with all the prior bindings and constraints. This will give you the type and boundness at the point of the star import definition (if unbound just return `None` and go on to the next binding.) The type will then replace `binding_ty` below, and possibly-unbound will mean the overall symbol resolution also has to be possibly-unbound.

The alternative would be to flatten all these prior bindings into the current loop somehow, with the visibility and narrowing constraints of the star-import binding ANDed with each one of the prior bindings. This is probably doable, but seems more complex?

Reply:
A test case to capture this would be

`module.py`:

```py
if False:
    X = 3
```

`main.py`:
```py
def flag() -> bool:
    return False

if flag():
    X = 1
else:
    X = 2

from module import *

reveal_type(X)  # Should be Literal[1, 2], currently results in Literal[1]
```

Reply:
This is a really interesting idea, and potentially much more elegant than the solution Carl and I had been working on! I'm going to try it out.